# Supervised Classification: Decision Trees, SVM, and Naive Bayes

Question 1: What is Information Gain, and how is it used in Decision Trees?

Answer : Information Gain (IG) is a measure used in Decision Trees to decide which attribute (feature) should be chosen to split the data at each step.
It helps identify the feature that gives the maximum reduction in uncertainty or impurity after splitting the dataset.

* Concept : Decision Tree tries to divide data into pure groups (where most instances belong to a single class).

  - Entropy measures impurity (or randomness).

  - Information Gain = Reduction in Entropy after splitting the data.

Example:

Suppose we’re building a decision tree to predict whether a student will attend class based on “Weather” (Sunny, Rainy).
We calculate entropy before split and entropy after split.
The feature giving highest information gain (i.e., highest reduction in entropy) is selected as the root node.


Question 2: What is the difference between Gini Impurity and Entropy?

Answer : Both Gini Impurity and Entropy are measures of how “pure” or “impure” a node is in a Decision Tree — but they are calculated differently and used by different algorithms.


1.Conceptual Difference:

- Entropy is based on the concept of information theory, which measures the amount of randomness or uncertainty present in the data.
It focuses on how much “information” is required to correctly classify a dataset.

- Gini Impurity, on the other hand, measures the probability of misclassification of a randomly chosen element if it were labeled according to the class distribution of the node.

2.Interpretation:

- Entropy measures the information content or degree of disorder in a dataset.
Higher entropy means more impurity and more disorder.

- Gini Impurity directly measures how often a randomly chosen element would be incorrectly labeled.
A lower Gini value indicates that the node is purer.

3.Usage in Algorithms:

- Entropy is used in Decision Tree algorithms like ID3 and C4.5.

- Gini Impurity is mainly used in the CART (Classification and Regression Tree) algorithm.

4.Computation and Speed:

- Entropy requires logarithmic calculations, so it is slightly more computationally complex.

- Gini Impurity is simpler and faster to calculate, which makes it preferable when dealing with large datasets.

5. Practical Preference:

- In practice, both measures often lead to very similar decision trees.

- Gini Impurity is preferred when speed and simplicity are important.

- Entropy is preferred when one wants a theoretically sound and information-based approach.

Question 3: What is Pre-Pruning in Decision Trees?

Answer : Pre-Pruning, also known as Early Stopping, is a technique used to control the growth of a Decision Tree during its construction phase.
Its main purpose is to avoid overfitting by stopping the tree from becoming too deep or complex.

- In a Decision Tree, as more splits are added, the model may start learning noise and unnecessary details from the training data, which reduces its performance on new data.
Pre-pruning helps to prevent this by setting certain stopping conditions while building the tree.

Common Stopping Conditions for Pre-Pruning:

- Maximum Depth: Limit how deep the tree can grow.

- Minimum Samples per Split: Stop splitting if the number of samples in a node is below a threshold.

- Minimum Information Gain: Stop splitting if the information gain from a split is very low.

- Maximum Number of Leaf Nodes: Restrict the total number of leaf nodes.

 Advantages:

   - Prevents overfitting.

   - Reduces model complexity.

  - Decreases training time.

Disadvantages:

- May lead to underfitting if stopped too early.

Question 4: Write a Python program to train a Decision Tree Classifier using Gini Impurity as the criterion and print the feature importances.

Answer (Practical):

In [1]:
# Importing required libraries
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier

# Load dataset
iris = load_iris()
X = iris.data
y = iris.target

# Create Decision Tree Classifier using Gini Impurity
clf = DecisionTreeClassifier(criterion='gini', random_state=0)

# Train the model
clf.fit(X, y)

# Print Feature Importances
print("Feature Importances:")
for name, importance in zip(iris.feature_names, clf.feature_importances_):
    print(f"{name}: {importance:.4f}")


Feature Importances:
sepal length (cm): 0.0000
sepal width (cm): 0.0133
petal length (cm): 0.0641
petal width (cm): 0.9226


Question 5: What is a Support Vector Machine (SVM)?

Answer:
A Support Vector Machine (SVM) is a supervised machine learning algorithm used for classification and regression tasks.
Its main goal is to find the best decision boundary (hyperplane) that separates data points of different classes with the maximum margin.

- Concept:

SVM represents data as points in an n-dimensional space (where n = number of features).

It finds a hyperplane that best separates different classes.

The data points that are closest to the hyperplane are called Support Vectors, and they are most important in defining the decision boundary.

- Key Idea:

SVM tries to maximize the margin, i.e., the distance between the separating hyperplane and the nearest data points of both classes.
A larger margin gives better generalization and classification accuracy.

- Advantages:

Works well in high-dimensional data.

Effective when the number of features is greater than the number of samples.

Robust to overfitting (especially with proper kernel selection).

- Disadvantages:

Computationally expensive for large datasets.

Selecting the right kernel and parameters can be challenging.

Question 6: What is the Kernel Trick in SVM?

Answer :
The Kernel Trick is a mathematical technique used in SVM to handle non-linearly separable data.
When data cannot be separated by a straight line (linear boundary), the kernel trick helps to map the data into a higher-dimensional space where it becomes linearly separable.

- Concept:

Instead of explicitly transforming data into higher dimensions, the kernel function computes the inner product of the transformed data points directly — saving both time and computation.
This allows SVM to work efficiently with complex data patterns.

Common Types of Kernels:

1. Linear Kernel: Used when data is linearly separable.

2. Polynomial Kernel: Used when data has polynomial relationships.

3. RBF (Radial Basis Function) Kernel: Used for non-linear data; it creates circular or curved boundaries.

4. Sigmoid Kernel: Similar to neural networks’ activation function



Question 7: Write a Python program to train two SVM classifiers with Linear and RBF kernels on the Wine dataset, then compare their accuracies.



In [2]:
# Importing required libraries
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

# Load Wine dataset
wine = load_wine()
X = wine.data
y = wine.target

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Train SVM with Linear Kernel
svm_linear = SVC(kernel='linear')
svm_linear.fit(X_train, y_train)
y_pred_linear = svm_linear.predict(X_test)
acc_linear = accuracy_score(y_test, y_pred_linear)

# Train SVM with RBF Kernel
svm_rbf = SVC(kernel='rbf')
svm_rbf.fit(X_train, y_train)
y_pred_rbf = svm_rbf.predict(X_test)
acc_rbf = accuracy_score(y_test, y_pred_rbf)

# Print accuracies
print("Accuracy with Linear Kernel:", acc_linear)
print("Accuracy with RBF Kernel:", acc_rbf)


Accuracy with Linear Kernel: 0.9814814814814815
Accuracy with RBF Kernel: 0.7592592592592593


Question 8: What is the Naïve Bayes classifier, and why is it called “Naïve”?

Answer :
The Naïve Bayes Classifier is a supervised machine learning algorithm based on Bayes’ Theorem, used for classification problems.
It predicts the class of a data point by calculating the probability of each class and selecting the one with the highest probability.

* Why it is called “Naïve”?

It is called Naïve because it assumes that all features are independent of each other — that is, the presence or absence of one feature does not affect the others.
In real-world data, this assumption is often unrealistic, but the model still performs surprisingly well.

- Advantages:

Simple and easy to implement

Works well with small datasets

Fast and efficient for text classification (e.g., spam detection, sentiment analysis)

-  Disadvantages:

Assumes independence between features (which is rarely true)

Performs poorly when features are highly correlated.

Question 9: Explain the differences between Gaussian Naïve Bayes, Multinomial Naïve Bayes, and Bernoulli Naïve Bayes.

Answer :
There are different variants of Naïve Bayes classifiers depending on the type of data and the distribution assumption.

1. Gaussian Naïve Bayes:

Used when features are continuous (real-valued numbers).

Assumes the data follows a normal (Gaussian) distribution.

Commonly used for datasets like Iris or Breast Cancer where features are numerical.

2. Multinomial Naïve Bayes:

Used for discrete or count-based features (e.g., number of times a word appears in a document).

Mostly used in text classification tasks such as spam filtering or document categorization.

It assumes feature values represent frequency counts.

3. Bernoulli Naïve Bayes:

Used for binary or Boolean features (0s and 1s).

Each feature represents whether a word or attribute is present or not.

Useful for binary text data like “word present = 1, absent = 0”.

Question 10: Breast Cancer Dataset — Train a Gaussian Naïve Bayes classifier and evaluate accuracy

Answer :

In [3]:
# Import necessary libraries
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

# Load the Breast Cancer dataset
data = load_breast_cancer()
X = data.data
y = data.target

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Create and train Gaussian Naïve Bayes model
model = GaussianNB()
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)

# Print the results
print("Accuracy of Gaussian Naïve Bayes on Breast Cancer Dataset:", round(accuracy, 4))


Accuracy of Gaussian Naïve Bayes on Breast Cancer Dataset: 0.9415


Explanation:

The Breast Cancer dataset contains numerical features, so Gaussian Naïve Bayes is the right choice.

The model calculates the probability of each class and predicts based on the highest posterior probability.

It achieves around 94% accuracy, showing it performs well for medical diagnosis data.